runtimes/pythonrt/svc.go

```diff
@@ -0,0 +1,360 @@
+	}
+	_, err := s.cbs.Call(s.ctx(), s.runID, s.syscallFn, args, nil)
+	if err != nil {
+		err = status.Errorf(codes.Internal, "subscribe(%s) -> %s", req.SignalId, err)
+		return &pb.UnsubscribeResponse{Error: err.Error()}, err
+	}
+
+	return &pb.UnsubscribeResponse{}, nil
+}
+
+func (s *remoteSvc) call(val sdktypes.Value) {
+	req := pb.ActivityReplyRequest{}
+
+	// We want to send reply in any case
+	defer func() {
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+		defer cancel()
+		reply, err := s.runner.ActivityReply(ctx, &req)
+		switch {
+		case err != nil:
+			s.log.Error("activity reply error", zap.Error(err))
+		case reply.Error != "":
+			s.log.Error("activity reply error", zap.String("error", reply.Error))
+		}
+	}()
+
+	if !val.IsFunction() {
+		s.log.Error("bad function", zap.Any("val", val))
+		req.Error = fmt.Sprintf("%#v is not a function", val)
+		return
+	}
+
+	fn := val.GetFunction()
+	req.Data = fn.Data()
+	out, err := s.cbs.Call(s.ctx(), s.runID, val, nil, nil)
+
+	switch {
+	case err != nil:
+		req.Error = fmt.Sprintf("%s - %s", fn.Name().String(), err)
+		s.log.Error("activity reply error", zap.Error(err))
+	case !out.IsBytes():
+		req.Error = fmt.Sprintf("call output not bytes: %#v", out)
+		s.log.Error("activity reply error", zap.String("error", req.Error))
+	default:
+		data := out.GetBytes().Value()
+		req.Result = data
+	}
+}
+
+// Runner starting activity
+func (s *remoteSvc) Activity(ctx context.Context, req *pb.ActivityRequest) (*pb.ActivityResponse, error) {
+	fnName := req.CallInfo.Function
+	s.log.Info("activity", zap.String("function", fnName))
+	fn, err := sdktypes.NewFunctionValue(s.xid, fnName, req.Data, nil, pyModuleFunc)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "new function value: %s", err)
+	}
+
+	go s.call(fn)
+
+	return &pb.ActivityResponse{}, nil
+}
+
+func (s *remoteSvc) Done(ctx context.Context, req *pb.DoneRequest) (*pb.DoneResponse, error) {
+	s.done <- req
+	return &pb.DoneResponse{}, nil
+}
+
+type Healther interface {
+	Health(ctx context.Context, in *pb.HealthRequest, opts ...grpc.CallOption) (*pb.HealthResponse, error)
+}
+
+func waitForServer(name string, h Healther, timeout time.Duration) error {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+	start := time.Now()
+	var req pb.HealthRequest
+
+	for time.Since(start) <= timeout {
+		resp, err := h.Health(ctx, &req)
+		if err != nil || resp.Error != "" {
+			time.Sleep(10 * time.Millisecond)
+			continue
+		}
+
+		return nil
+	}
+
+	return fmt.Errorf("%s not ready after %v", name, timeout)
+}
+
+func freePort() (int, error) {
+	conn, err := net.Listen("tcp", ":0")
+	if err != nil {
+		return 0, err
+	}
+
+	conn.Close()
+	return conn.Addr().(*net.TCPAddr).Port, nil
+}
+
+// Start starts the server on a free port in a new goroutine.
+// It returns the port the server listens on.
+func (s *remoteSvc) Start() error {
+	port, err := freePort()
```

Comment:
not sure you need this freePort call

just use 
```go
lis, err := net.Listen("tcp", addr)
if err != nil {
 	return err
 }
s.lis = lis
s.port = lis.Addr().(*net.TCPAddr).Port
```

Reply:
do we really need to start a new server for each python process ?

we can add this capability to ak itself, and use the runner ID to get back to the correct go routine of the workflow itself

should be less resources, less error prone (I think)